File: ruby/lib/datagram_channel.rb
# This code is free software; you can redistribute it and/or modify it under the
# terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'ipaddr'
require 'socket'

require 'stringio_additions'

class DatagramChannel

  attr_reader :socket

  def self.open
    DatagramChannel.new
  end

  def close
    @socket.close
  end

  def connect(*args)
    if args[0].is_a? IPAddr
      ip_address = args[0].to_s
    elsif args[0].is_a? String
      ip_address = IPSocket.getaddress args[0]
    else
      raise ArgumentError
    end

    if args[1].is_a? Numeric
      port_number = args[1]
    else
      port_number = args[1].to_i
    end

    @socket.connect ip_address, port_number

    self
  end

  def read(destination_io)
    raise ArgumentError unless destination_io.is_a? StringIO

    data = @socket.recv destination_io.remaining
    length = destination_io.write data
    destination_io.truncate length
  end

  def write(source_io)
    raise ArgumentError unless source_io.is_a? StringIO

    @socket.send(source_io.get, 0)
  end

  protected

  def initialize
    @socket = UDPSocket.new
  end

end

File: php/lib/DatagramChannel.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2008-2009, Sebastian Staudt
 *
 * @author Sebastian Staudt
 * @license http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package Steam Condenser (PHP)
 * @subpackage DatagramChannel
 */

require_once "ByteBuffer.php";
require_once "InetAddress.php";
require_once "UDPSocket.php";

/**
 * @package Steam Condenser (PHP)
 * @subpackage DatagramChannel
 */
class DatagramChannel
{
	/**
	 * @var Socket
	 */
	private $socket;

	protected function __construct()
	{
		$this->socket = new UDPSocket();
		$this->configureBlocking(true);
	}

	public static function open()
	{
		return new DatagramChannel();
	}

	public function close()
	{
		$this->socket->close();
	}

	public function connect(InetAddress $ipAddress, $portNumber)
	{
		$this->socket->connect($ipAddress, $portNumber);
	}

	public function configureBlocking($doBlock)
	{
		$this->socket->setBlock($doBlock);
	}

	public function read(ByteBuffer $destinationBuffer)
	{
		$length = $destinationBuffer->remaining();
		$data = $this->socket->recv($length);

		$destinationBuffer->put($data);

		return strlen($data);
	}

	public function socket()
	{
		return $this->socket;
	}

	public function write(ByteBuffer $sourceBuffer)
	{
		return $this->socket->send($sourceBuffer->get());
	}
}
?>